Importing file with alias path is not working

# Description

Currently, the `kysely.config.ts` file is importing the `kysely` instance I created on a separate file and the path is an alias path like `@/path/to/kysely`.

This throws an error on the CLI.

Looking at `c12`, it looks like it's exposing [`jitiOptions`](https://github.com/unjs/c12?tab=readme-ov-file#jitioptions) which also exposes the [`alias`](https://github.com/unjs/jiti?tab=readme-ov-file#alias) to allow alias path.

```
ERROR  Cannot find module '@/path/to/kysely'
Require stack:
- /repo/kysely.config.ts

  Require stack:
  - kysely.config.ts
  at Module._resolveFilename (node:internal/modules/cjs/loader:1143:15)
  at Function.resolve (node:internal/modules/helpers:190:19)
  at _resolve (node_modules/jiti/dist/jiti.js:1:241814)
  at jiti (node_modules/jiti/dist/jiti.js:1:244531)
  at kysely.config.ts:2:15
  at evalModule (node_modules/jiti/dist/jiti.js:1:247313)
  at Object.jiti (node_modules/jiti/dist/jiti.js:1:245241)
  at resolveConfig (node_modules/c12/dist/shared/c12.cab0c9da.mjs:345:26)
  at loadConfig (node_modules/c12/dist/shared/c12.cab0c9da.mjs:147:29)
  at async getConfig (node_modules/kysely-ctl/dist/bin.js:226:24)
```

# Workaround

Manually configure using `dialect` and `dialectConfig` essentially repeating the same logic.

# Package Info

```
kysely v0.27.4
kysely-ctl v0.9.0
```

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -58,7 +58,6 @@ jobs:
           cache: "pnpm"
 
       - name: Calculate additional flags
-        if: matrix.type == 'esm-tsconfig-paths'
         id: flags
         shell: bash
         run: |

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -57,6 +57,11 @@ jobs:
           node-version: ${{ matrix.node-version }}
           cache: "pnpm"
 
+      - name: Calculate additional flags
+        if: matrix.type == 'esm-tsconfig-paths'
+        id: flags
+        run: echo "ADDITIONAL_FLAGS=--experimental-resolve-tsconfig-paths" >> $GITHUB_OUTPUT
+
       - name: Install dependencies
         run: pnpm i && pnpm build && cd examples/node-${{ matrix.type }} && pnpm i
 
@@ -67,24 +72,24 @@ jobs:
       - name: kysely migrate:make <migration-name> --cwd=%cd%
         if: matrix.os == 'windows-latest'
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:make moshe --cwd=%cd% --debug
+        run: pnpm kysely migrate:make moshe --cwd=%cd% --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:make <migration-name>
         if: matrix.os != 'windows-latest'
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:make moshe --debug
+        run: pnpm kysely migrate:make moshe --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:list
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:list --debug
+        run: pnpm kysely migrate:list --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:latest
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:latest --debug
+        run: pnpm kysely migrate:latest --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:rollback --all
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:rollback --all --debug
+        run: pnpm kysely migrate:rollback --all --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
   bun:
     runs-on: ubuntu-latest

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -72,24 +72,24 @@ jobs:
       - name: kysely migrate:make <migration-name> --cwd=%cd%
         if: matrix.os == 'windows-latest'
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:make moshe --cwd=%cd% --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:make moshe --cwd=%cd% --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:make <migration-name>
         if: matrix.os != 'windows-latest'
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:make moshe --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:make moshe --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:list
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:list --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:list --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:latest
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:latest --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:latest --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:rollback --all
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:rollback --all --debug {{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:rollback --all --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
   bun:
     runs-on: ubuntu-latest

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -60,7 +60,7 @@ jobs:
       - name: Calculate additional flags
         if: matrix.type == 'esm-tsconfig-paths'
         id: flags
-        run: echo "ADDITIONAL_FLAGS=--experimental-resolve-tsconfig-paths" >> "$GITHUB_OUTPUT"
+        run: echo "ADDITIONAL_FLAGS='--experimental-resolve-tsconfig-paths'" >> "$GITHUB_OUTPUT"
 
       - name: Install dependencies
         run: pnpm i && pnpm build && cd examples/node-${{ matrix.type }} && pnpm i

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -38,7 +38,7 @@ jobs:
       matrix:
         node-version: [18, 20, 22, 23]
         os: [macos-latest, ubuntu-latest, windows-latest]
-        type: [cjs, esm]
+        type: [cjs, esm, esm-tsconfig-paths]
       fail-fast: false
 
     runs-on: ${{ matrix.os }}

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -80,19 +80,19 @@ jobs:
         working-directory: examples/node-${{ matrix.type }}
         run: pnpm kysely -v
 
-      - name: kysely migrate:make <migration-name>
+      - name: kysely migrate:make <migration-name> ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
         working-directory: examples/node-${{ matrix.type }}
         run: pnpm kysely migrate:make moshe ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
-      - name: kysely migrate:list
+      - name: kysely migrate:list ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
         working-directory: examples/node-${{ matrix.type }}
         run: pnpm kysely migrate:list ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
-      - name: kysely migrate:latest
+      - name: kysely migrate:latest ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
         working-directory: examples/node-${{ matrix.type }}
         run: pnpm kysely migrate:latest ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
-      - name: kysely migrate:rollback --all
+      - name: kysely migrate:rollback --all ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
         working-directory: examples/node-${{ matrix.type }}
         run: pnpm kysely migrate:rollback --all ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -61,7 +61,18 @@ jobs:
         if: matrix.type == 'esm-tsconfig-paths'
         id: flags
         shell: bash
-        run: echo "ADDITIONAL_FLAGS='--experimental-resolve-tsconfig-paths'" >> "$GITHUB_OUTPUT"
+        run: |
+          ADDITIONAL_FLAGS="--debug"
+
+          if [[ ${{ matrix.type }} == 'esm-tsconfig-paths' ]]; then
+            ADDITIONAL_FLAGS+=" --experimental-resolve-tsconfig-paths"
+          fi
+          
+          if [[ ${{ matrix.os }} == 'windows-latest' ]]; then
+            ADDITIONAL_FLAGS+=" --cwd=%cd%"
+          fi
+
+          echo "ADDITIONAL_FLAGS=$ADDITIONAL_FLAGS" >> $GITHUB_OUTPUT
 
       - name: Install dependencies
         run: pnpm i && pnpm build && cd examples/node-${{ matrix.type }} && pnpm i
@@ -70,27 +81,21 @@ jobs:
         working-directory: examples/node-${{ matrix.type }}
         run: pnpm kysely -v
 
-      - name: kysely migrate:make <migration-name> --cwd=%cd%
-        if: matrix.os == 'windows-latest'
-        working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:make moshe --cwd=%cd% --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
-
       - name: kysely migrate:make <migration-name>
-        if: matrix.os != 'windows-latest'
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:make moshe --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:make moshe ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:list
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:list --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:list ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:latest
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:latest --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:latest ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
       - name: kysely migrate:rollback --all
         working-directory: examples/node-${{ matrix.type }}
-        run: pnpm kysely migrate:rollback --all --debug ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
+        run: pnpm kysely migrate:rollback --all ${{ steps.flags.outputs.ADDITIONAL_FLAGS }}
 
   bun:
     runs-on: ubuntu-latest

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -60,6 +60,7 @@ jobs:
       - name: Calculate additional flags
         if: matrix.type == 'esm-tsconfig-paths'
         id: flags
+        shell: bash
         run: echo "ADDITIONAL_FLAGS='--experimental-resolve-tsconfig-paths'" >> "$GITHUB_OUTPUT"
 
       - name: Install dependencies

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -60,7 +60,7 @@ jobs:
       - name: Calculate additional flags
         if: matrix.type == 'esm-tsconfig-paths'
         id: flags
-        run: echo "ADDITIONAL_FLAGS=--experimental-resolve-tsconfig-paths" >> $GITHUB_OUTPUT
+        run: echo "ADDITIONAL_FLAGS=--experimental-resolve-tsconfig-paths" >> "$GITHUB_OUTPUT"
 
       - name: Install dependencies
         run: pnpm i && pnpm build && cd examples/node-${{ matrix.type }} && pnpm i

--- a/examples/node-esm-tsconfig-paths/.config/kysely.config.ts
+++ b/examples/node-esm-tsconfig-paths/.config/kysely.config.ts
@@ -1,0 +1,7 @@
+import { defineConfig } from 'kysely-ctl'
+
+import kysely from '@/kysely'
+
+export default defineConfig({
+	kysely,
+})

--- a/examples/node-esm-tsconfig-paths/migrations/1716743937856_add_table_moshe.ts
+++ b/examples/node-esm-tsconfig-paths/migrations/1716743937856_add_table_moshe.ts
@@ -1,0 +1,10 @@
+import { MOSHE_TABLE } from '@/config/db'
+import type { Kysely } from 'kysely'
+
+export async function up(db: Kysely<any>): Promise<void> {
+	await db.schema.createTable(MOSHE_TABLE).addColumn('id', 'integer').execute()
+}
+
+export async function down(db: Kysely<any>): Promise<void> {
+	await db.schema.dropTable(MOSHE_TABLE).execute()
+}

--- a/examples/node-esm-tsconfig-paths/migrations/1716746051776_add_column_is_moshe.ts
+++ b/examples/node-esm-tsconfig-paths/migrations/1716746051776_add_column_is_moshe.ts
@@ -1,0 +1,13 @@
+import { MOSHE_TABLE } from '@/config/db'
+import type { Kysely } from 'kysely'
+
+export async function up(db: Kysely<any>): Promise<void> {
+	await db.schema
+		.alterTable(MOSHE_TABLE)
+		.addColumn('is_moshe', 'boolean')
+		.execute()
+}
+
+export async function down(db: Kysely<any>): Promise<void> {
+	await db.schema.alterTable(MOSHE_TABLE).dropColumn('is_moshe').execute()
+}

--- a/examples/node-esm-tsconfig-paths/package.json
+++ b/examples/node-esm-tsconfig-paths/package.json
@@ -1,0 +1,12 @@
+{
+	"name": "node-esm",
+	"type": "module",
+	"devDependencies": {
+		"@types/better-sqlite3": "^7.6.12",
+		"kysely-ctl": "link:../.."
+	},
+	"dependencies": {
+		"better-sqlite3": "^11.8.1",
+		"kysely": "^0.27.5"
+	}
+}

--- a/examples/node-esm-tsconfig-paths/src/config/db.ts
+++ b/examples/node-esm-tsconfig-paths/src/config/db.ts
@@ -1,0 +1,1 @@
+export const DB_PATH = './example.db'

--- a/examples/node-esm-tsconfig-paths/src/config/db.ts
+++ b/examples/node-esm-tsconfig-paths/src/config/db.ts
@@ -1,1 +1,3 @@
 export const DB_PATH = './example.db'
+
+export const MOSHE_TABLE = 'moshe'

--- a/examples/node-esm-tsconfig-paths/src/kysely.ts
+++ b/examples/node-esm-tsconfig-paths/src/kysely.ts
@@ -1,7 +1,7 @@
 import database from 'better-sqlite3'
 import { Kysely, SqliteDialect } from 'kysely'
 
-import { DB_PATH } from '@config/db'
+import { DB_PATH } from '@/config/db'
 
 export default new Kysely({
 	dialect: new SqliteDialect({

--- a/examples/node-esm-tsconfig-paths/src/kysely.ts
+++ b/examples/node-esm-tsconfig-paths/src/kysely.ts
@@ -1,0 +1,10 @@
+import database from 'better-sqlite3'
+import { Kysely, SqliteDialect } from 'kysely'
+
+import { DB_PATH } from '@config/db'
+
+export default new Kysely({
+	dialect: new SqliteDialect({
+		database: database(DB_PATH),
+	}),
+})

--- a/examples/node-esm-tsconfig-paths/tsconfig.json
+++ b/examples/node-esm-tsconfig-paths/tsconfig.json
@@ -1,0 +1,11 @@
+{
+	"compilerOptions": {
+		"baseUrl": "./",
+		"paths": {
+			"@/*": ["src/*"],
+			"@config/*": ["src/config/*"]
+		}
+	},
+	"extends": "../../tsconfig.base.json",
+	"include": [".config/**/*", "src", "migrations"]
+}

--- a/examples/node-esm-tsconfig-paths/tsconfig.json
+++ b/examples/node-esm-tsconfig-paths/tsconfig.json
@@ -2,8 +2,7 @@
 	"compilerOptions": {
 		"baseUrl": "./",
 		"paths": {
-			"@/*": ["src/*"],
-			"@config/*": ["src/config/*"]
+		 	"@/*": ["./src/*"]
 		}
 	},
 	"extends": "../../tsconfig.base.json",

--- a/src/arguments/jiti.mts
+++ b/src/arguments/jiti.mts
@@ -1,5 +1,14 @@
 import type { ArgsDef } from 'citty'
 
+const ExperimentalResolveTSConfigPathsArg = {
+	'experimental-resolve-tsconfig-paths': {
+		default: false,
+		description:
+			'Attempts to resolve path aliases using the tsconfig.json file.',
+		type: 'boolean',
+	},
+} satisfies ArgsDef
+
 const NoFilesystemCachingArg = {
 	'no-filesystem-caching': {
 		default: false,
@@ -10,5 +19,6 @@ const NoFilesystemCachingArg = {
 } satisfies ArgsDef
 
 export const JitiArgs = {
+	...ExperimentalResolveTSConfigPathsArg,
 	...NoFilesystemCachingArg,
 } satisfies ArgsDef

--- a/src/config/get-config.mts
+++ b/src/config/get-config.mts
@@ -12,6 +12,7 @@ export interface ArgsLike {
 	cwd?: string
 	debug?: boolean
 	environment?: string
+	'experimental-resolve-tsconfig-paths'?: boolean
 	'filesystem-caching'?: boolean
 }
 
@@ -25,7 +26,9 @@ export async function getConfig(
 		dotenv: true,
 		envName: args.environment,
 		jitiOptions: {
-			alias: await getJitiAlias(),
+			alias: args['experimental-resolve-tsconfig-paths']
+				? await getJitiAlias()
+				: undefined,
 			debug: Boolean(args.debug),
 			fsCache: Boolean(args['filesystem-caching']),
 		},

--- a/src/config/get-config.mts
+++ b/src/config/get-config.mts
@@ -1,5 +1,6 @@
 import { loadConfig } from 'c12'
 import { consola } from 'consola'
+import { getJitiAlias } from '../utils/jiti'
 import { getCWD } from './get-cwd.mjs'
 import { getMillisPrefix } from './get-file-prefix.mjs'
 import type {
@@ -24,6 +25,7 @@ export async function getConfig(
 		dotenv: true,
 		envName: args.environment,
 		jitiOptions: {
+			alias: await getJitiAlias(),
 			debug: Boolean(args.debug),
 			fsCache: Boolean(args['filesystem-caching']),
 		},

--- a/src/utils/jiti.ts
+++ b/src/utils/jiti.ts
@@ -1,0 +1,45 @@
+import { join } from 'pathe'
+import type { CompilerOptions } from 'typescript'
+import { getCWD } from '../config/get-cwd.mjs'
+import { getTSConfig } from './tsconfig.mjs'
+
+export interface ArgsLike {
+	cwd?: string
+}
+
+export async function getJitiAlias(): Promise<
+	Record<string, string> | undefined
+> {
+	try {
+		const tsconfig = await getTSConfig()
+
+		const { baseUrl, paths } = (tsconfig.compilerOptions ||
+			{}) as CompilerOptions
+
+		if (!paths) {
+			return
+		}
+
+		const cwd = getCWD()
+
+		const jitiAlias: Record<string, string> = {}
+
+		for (const [alias, [path]] of Object.entries(paths)) {
+			if (!path) {
+				continue
+			}
+
+			jitiAlias[removeWildcards(alias)] = removeWildcards(
+				join(cwd, baseUrl || '.', path),
+			)
+		}
+
+		return jitiAlias
+	} catch (error) {
+		return {}
+	}
+}
+
+function removeWildcards(path: string): string {
+	return path.replace(/(\/?\*\*?)+$/, '')
+}

--- a/src/utils/jiti.ts
+++ b/src/utils/jiti.ts
@@ -3,10 +3,6 @@ import type { CompilerOptions } from 'typescript'
 import { getCWD } from '../config/get-cwd.mjs'
 import { getTSConfig } from './tsconfig.mjs'
 
-export interface ArgsLike {
-	cwd?: string
-}
-
 export async function getJitiAlias(): Promise<
 	Record<string, string> | undefined
 > {

--- a/src/utils/tsconfig.mts
+++ b/src/utils/tsconfig.mts
@@ -1,0 +1,10 @@
+import { type TSConfig, readTSConfig } from 'pkg-types'
+import { getCWD } from '../config/get-cwd.mjs'
+
+let tsconfig: TSConfig
+
+export async function getTSConfig(): Promise<TSConfig> {
+	return (tsconfig ||= await readTSConfig(undefined, {
+		startingFrom: getCWD(),
+	}))
+}

